File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_cpp/p03575/p03575_num3_parsed.cpp
#include <iostream>
#include <vector>
using namespace std;

void dfs(int node, int parent, int& timer, vector<int>& disc, vector<int>& low, const vector<vector<int>>& adj, vector<pair<int, int>>& bridges) {
    disc[node] = low[node] = timer++;
    
    for (int neighbor : adj[node]) {
        if (disc[neighbor] == -1) {
            dfs(neighbor, node, timer, disc, low, adj, bridges);
            low[node] = min(low[node], low[neighbor]);
            
            if (low[neighbor] > disc[node]) {
                bridges.push_back({node, neighbor});
            }
        } else if (neighbor != parent) {
            low[node] = min(low[node], disc[neighbor]);
        }
    }
}

int findBridges(int n, const vector<pair<int, int>>& edges) {
    vector<vector<int>> adj(n + 1);
    
    for (auto edge : edges) {
        int u = edge.first;
        int v = edge.second;
        adj[u].push_back(v);
        adj[v].push_back(u);
    }
    
    int timer = 0;
    vector<int> disc(n + 1, -1);
    vector<int> low(n + 1);
    vector<pair<int, int>> bridges;
    
    dfs(1, -1, timer, disc, low, adj, bridges);
    
    return bridges.size();
}

int main() {
    int n, m;
    cin >> n >> m;
    
    vector<pair<int, int>> edges(m);
    for (int i = 0; i < m; i++) {
        cin >> edges[i].first >> edges[i].second;
    }
    
    int numBridges = findBridges(n, edges);
    
    cout << numBridges << endl;
    
    return 0;
}

File: Datasets/TrickyBugs/GenProgs/dpp_generated_progs_python/p03575/p03575_num3_parsed.py
def find_bridges(graph, n, m):
    def dfs(v, p):
        nonlocal timer

        low[v] = ids[v] = timer
        timer += 1

        for u in graph[v]:
            if u == p:
                continue
            if ids[u] == -1:
                dfs(u, v)
                low[v] = min(low[v], low[u])
                if low[u] > ids[v]:
                    bridges.add((v, u))
            else:
                low[v] = min(low[v], ids[u])

    timer = 0
    ids = [-1] * (n + 1)
    low = [0] * (n + 1)
    bridges = set()

    for i in range(1, n + 1):
        if ids[i] == -1:
            dfs(i, -1)

    return bridges


def main():
    n, m = map(int, input().split())
    graph = {i: [] for i in range(1, n + 1)}

    for _ in range(m):
        a, b = map(int, input().split())
        graph[a].append(b)
        graph[b].append(a)

    bridges = find_bridges(graph, n, m)

    print(len(bridges))


if __name__ == "__main__":
    main()